# Import Essential Libraries


In [29]:
# Basic Libraries
import numpy as np
import pandas as pd


# Extract Data in CSV file

In [30]:
df = pd.read_csv('loan_approval_dataset.csv')
df.head()

loan_id   no_of_dependents      education  self_employed   income_annum  \
0        1                  2       Graduate             No        9600000   
1        2                  0   Not Graduate            Yes        4100000   
2        3                  3       Graduate             No        9100000   
3        4                  3       Graduate             No        8200000   
4        5                  5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score   residential_assets_value  \
0      29900000          12           778                    2400000   
1      12200000           8           417                    2700000   
2      29700000          20           506                    7100000   
3      30700000           8           467                   18200000   
4      24200000          20           382                   12400000   

    commercial_assets_value   luxury_assets_value   bank_asset_value  \
0                  17600000              22700000            8000000   
1                   2200000               8800000            3300000   
2                   4500000              33300000           12800000   
3                   3300000              23300000            7900000   
4                   8200000              29400000            5000000   

   loan_status  
0     Approved  
1     Rejected  
2     Rejected  
3     Rejected  
4     Rejected

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4269 entries, 0 to 4268
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   loan_id                    4269 non-null   int64 
 1    no_of_dependents          4269 non-null   int64 
 2    education                 4269 non-null   object
 3    self_employed             4269 non-null   object
 4    income_annum              4269 non-null   int64 
 5    loan_amount               4269 non-null   int64 
 6    loan_term                 4269 non-null   int64 
 7    cibil_score               4269 non-null   int64 
 8    residential_assets_value  4269 non-null   int64 
 9    commercial_assets_value   4269 non-null   int64 
 10   luxury_assets_value       4269 non-null   int64 
 11   bank_asset_value          4269 non-null   int64 
 12   loan_status               4269 non-null   object
dtypes: int64(10), object(3)
memory usage: 433.7+ KB


# Data Cleaning and Preproccessing

##### Check to see if there is any missing values or duplicates in dataset.

In [32]:
# check for missing values
df.isnull().sum()

loan_id                      0
 no_of_dependents            0
 education                   0
 self_employed               0
 income_annum                0
 loan_amount                 0
 loan_term                   0
 cibil_score                 0
 residential_assets_value    0
 commercial_assets_value     0
 luxury_assets_value         0
 bank_asset_value            0
 loan_status                 0
dtype: int64

In [33]:
#check for duplicate values
duplicate_rows = df.duplicated()
duplicates = duplicate_rows.sum()
print("Total duplicates:", duplicates)

Total duplicates: 0


> There is **no missing data or duplicates** in dataset.

#### Dropping unneccessary columns and whitespaces in dataset 

In [34]:
# Dropping Unnecessary Columns
df = df.drop(["loan_id"],axis=1)
df.head()

no_of_dependents      education  self_employed   income_annum  \
0                  2       Graduate             No        9600000   
1                  0   Not Graduate            Yes        4100000   
2                  3       Graduate             No        9100000   
3                  3       Graduate             No        8200000   
4                  5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score   residential_assets_value  \
0      29900000          12           778                    2400000   
1      12200000           8           417                    2700000   
2      29700000          20           506                    7100000   
3      30700000           8           467                   18200000   
4      24200000          20           382                   12400000   

    commercial_assets_value   luxury_assets_value   bank_asset_value  \
0                  17600000              22700000            8000000   
1                   2200000               8800000            3300000   
2                   4500000              33300000           12800000   
3                   3300000              23300000            7900000   
4                   8200000              29400000            5000000   

   loan_status  
0     Approved  
1     Rejected  
2     Rejected  
3     Rejected  
4     Rejected

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4269 entries, 0 to 4268
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0    no_of_dependents          4269 non-null   int64 
 1    education                 4269 non-null   object
 2    self_employed             4269 non-null   object
 3    income_annum              4269 non-null   int64 
 4    loan_amount               4269 non-null   int64 
 5    loan_term                 4269 non-null   int64 
 6    cibil_score               4269 non-null   int64 
 7    residential_assets_value  4269 non-null   int64 
 8    commercial_assets_value   4269 non-null   int64 
 9    luxury_assets_value       4269 non-null   int64 
 10   bank_asset_value          4269 non-null   int64 
 11   loan_status               4269 non-null   object
dtypes: int64(9), object(3)
memory usage: 400.3+ KB


 > Unneccessary variable `loan_id` removed from dataset

From the **new dataframe**, it could be observed that :

- There are **12 variables** and **4269 samples** in the dataset

- The **response** variable is the `loan_status`

Classifying them based on `Dtype`:

- There are **9 variables** identified as 'int64' by default. 

- There are **2 variables** identified as 'object' by default and they are categorical.


In [36]:
unique_val = df[' no_of_dependents'].unique()
no_unique_val = df[' no_of_dependents'].nunique()
print("unique values for no_of_dependents:", unique_val)
print("Number of unique values for no_of_dependents:", no_unique_val)
df[' no_of_dependents'] = df[' no_of_dependents'].astype('object')


unique values for no_of_dependents: [2 0 3 5 4 1]
Number of unique values for no_of_dependents: 6


> Even though, `no_of_dependents` is classified under int64, it is a **categorical** as it is bound between (0,6) which is a limited range of values.

# Creating new variables

In the new dataframe, there are 4 types of assets : `residential_assets_value` , `commercial_assets_value` , `luxury_assets_value` , `bank_asset_value`.

We would use the process of **feature engineering** by aggregating the features stated into 2 types of assets : `movable_assets` and `immovable_assets`

> `residential_assets_value` , `commercial_assets_value` will be grouped into     `immovable_assets`

>  `luxury_assets_value` , `bank_asset_value` will be grouped into `movable_assets`

Additionally, `loan_to_income`and `assets_to_loan` ratio could be added as new features.

`loan_to_income` measures how large the requested loan amount is relative to the applicant's income. Indicates the debt burden that the loan would place on the borrower.

> Higher value might indicate higher risk of default, while lower ratio suggests that borrower might manage repayments more comfortably. Helps model to understand repayment capacity from a different perspective.

`assets_to_loan` measures the total assets a borrower has in relation to the loan amount they are requesting. It reflects potential collateral that could be used in case of a default.

> Provides insights into the collateral value relative to the loan. Higher asset coverage might suggest lower risk for the lender, improving chances of approval.



Results of grouping existing variables :

- **Reduction in multicollinearity**, potential improvement in model performance.

- **Highlight new relationships**: New variables created might better highlight correlation between assets and response variable by capturing combined effects that individual assets values do not.

- Reduction in number of variables, **greater simplification of data** used for models.


Moreover,
`cibil_score` also known as credit score, is an important aspect that 
financial institution consider when determining loan eligibility. We will be
using the Cibil scoring system provided by TransUnion Credit Information Bureau (India) Limited to **categorize** the cibil scores into 4 different
categories. By doing this, we can give more meaning to the data and understand the creditworthiness of each category. 

> Instead of using `cibil_score` as a numerical value (int64), we would create a new variable `cibil_category`(category) for **better analysis**.


In [37]:
# Movable Assets
df[' movable_assets'] = df[' bank_asset_value'] + df[' luxury_assets_value']

# Immovable Assets
df[' immovable_assets'] = df[' residential_assets_value'] + df[' commercial_assets_value']

# Drop Columns
df.drop(columns=[' bank_asset_value',' luxury_assets_value', ' residential_assets_value', ' commercial_assets_value' ], inplace=True)

df['loan_to_income_ratio'] = df[' loan_amount'] / df[' income_annum']
df['assets_to_loan_ratio'] = (df[' movable_assets'] + df[' immovable_assets']) / df[' loan_amount']
df.head()

no_of_dependents      education  self_employed   income_annum  \
0                 2       Graduate             No        9600000   
1                 0   Not Graduate            Yes        4100000   
2                 3       Graduate             No        9100000   
3                 3       Graduate             No        8200000   
4                 5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score  loan_status   movable_assets  \
0      29900000          12           778     Approved         30700000   
1      12200000           8           417     Rejected         12100000   
2      29700000          20           506     Rejected         46100000   
3      30700000           8           467     Rejected         31200000   
4      24200000          20           382     Rejected         34400000   

    immovable_assets  loan_to_income_ratio  assets_to_loan_ratio  
0           20000000              3.114583              1.695652  
1            4900000              2.975610              1.393443  
2           11600000              3.263736              1.942761  
3           21500000              3.743902              1.716612  
4           20600000              2.469388              2.272727

In [38]:
# Define the boundaries for the categories
bins = [300, 550, 650, 750, 900]  # Adjust these as per your cibil score system
labels = ['Poor', 'Fair', 'Good', 'Excellent']

# Categorize the 'cibil_score' without cutting from the original data
df[' cibil_category'] = pd.cut(df[' cibil_score'], bins=bins, labels=labels, include_lowest=True)


In [39]:
df.describe().T.style.background_gradient(cmap='Blues')

In [40]:
df.head()

no_of_dependents      education  self_employed   income_annum  \
0                 2       Graduate             No        9600000   
1                 0   Not Graduate            Yes        4100000   
2                 3       Graduate             No        9100000   
3                 3       Graduate             No        8200000   
4                 5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score  loan_status   movable_assets  \
0      29900000          12           778     Approved         30700000   
1      12200000           8           417     Rejected         12100000   
2      29700000          20           506     Rejected         46100000   
3      30700000           8           467     Rejected         31200000   
4      24200000          20           382     Rejected         34400000   

    immovable_assets  loan_to_income_ratio  assets_to_loan_ratio  \
0           20000000              3.114583              1.695652   
1            4900000              2.975610              1.393443   
2           11600000              3.263736              1.942761   
3           21500000              3.743902              1.716612   
4           20600000              2.469388              2.272727   

   cibil_category  
0       Excellent  
1            Poor  
2            Poor  
3            Poor  
4            Poor

In [41]:

df.columns = df.columns.str.strip()


df.columns


Index(['no_of_dependents', 'education', 'self_employed', 'income_annum',
       'loan_amount', 'loan_term', 'cibil_score', 'loan_status',
       'movable_assets', 'immovable_assets', 'loan_to_income_ratio',
       'assets_to_loan_ratio', 'cibil_category'],
      dtype='object')

In [42]:
df.to_csv('cleaned_loan_approval_dataset.csv', index = False)